src/Pact/Types/Command.hs

```diff
@@ -120,56 +121,83 @@ requestKeyHash :: H.Blake2b_512
+             m ->
+             Text ->
+             PactRPC c ->
+             IO (Command ByteString)
+mkCommand creds meta nonce rpc = mkCommand' creds encodedPayload
+  where encodedPayload = BSL.toStrict $ A.encode payload
+        payload = Payload rpc nonce meta $ toSigners creds
+
+toSigners :: [SomeKeyPair] -> [Signer]
+toSigners creds = map toSigner creds
+  where toSigner cred = Signer
+                        (kpToPPKScheme cred)
+                        (toB16Text $ getPublic cred)
+                        (toB16Text $ formatPublicKey cred)
 
 
 mkCommand' :: [SomeKeyPair] -> ByteString -> IO (Command ByteString)
-mkCommand' creds env = makeCommand <$> (traverse makeSigs creds)
-  where makeCommand sigs = Command env sigs hsh
-        hsh = hashTx env requestKeyHash    -- hash associated with a Command, aka a Command's Request Key
-        makeSigs kp = mkUserSig hsh kp
+mkCommand' creds env = do
+  let hsh = hashTx env requestKeyHash    -- hash associated with a Command, aka a Command's Request Key
+      toUserSig cred = UserSig <$>
+                       toB16Text <$>
+                       sign cred hsh
+  sigs <- traverse toUserSig creds
+  return $ Command env sigs hsh
 
-mkUserSig :: Hash -> SomeKeyPair -> IO UserSig
-mkUserSig hsh kp =
-  let pub = toB16Text $ getPublic kp
-      formattedPub = toB16Text $ formatPublicKey kp
-      sig = toB16Text <$> sign kp hsh
-  in UserSig (kpToPPKScheme kp) pub formattedPub <$> sig
 
 
+-- VALIDATING TRANSACTIONS
 
 verifyCommand :: FromJSON m => Command ByteString -> ProcessedCommand m ParsedCode
-verifyCommand orig@Command{..} = case (ppcmdPayload', ppcmdHash', mSigIssue) of
-      (Right env', Right _, Nothing) -> ProcSucc $ orig { _cmdPayload = env' }
-      (e, h, s) -> ProcFail $ "Invalid command: " ++ toErrStr e ++ toErrStr h ++ fromMaybe "" s
+verifyCommand orig@Command{..} =
+  case isValidPayload of
+    Right env' -> case isValidPayloadHash of
+      Right hsh -> case (hasInvalidSigs hsh _cmdSigs $ _pSigners env') of
+        Nothing -> toProcSucc env'
+        Just sigErr -> toProcFail sigErr
+      Left hshErr -> toProcFail hshErr
+    Left payloadErr -> toProcFail payloadErr
   where
-    ppcmdPayload' = traverse parsePact =<< A.eitherDecodeStrict' _cmdPayload
+    toProcSucc parsedPayload = ProcSucc $ orig { _cmdPayload = parsedPayload }
+    toProcFail errStr = ProcFail $ "Invalid command: " ++ errStr
+
     parsePact :: Text -> Either String ParsedCode
     parsePact code = ParsedCode code <$> parseExprs code
-    (ppcmdSigs' :: [(UserSig,Bool)]) = (\u -> (u,verifyUserSig _cmdHash u)) <$> _cmdSigs
-    ppcmdHash' = verifyHashTx _cmdHash _cmdPayload requestKeyHash
-    mSigIssue = if all snd ppcmdSigs' then Nothing
-      else Just $ "Invalid sig(s) found: " ++ show (A.encode . fst <$> filter (not.snd) ppcmdSigs')
-    toErrStr :: Either String a -> String
-    toErrStr (Right _) = ""
-    toErrStr (Left s) = s ++ "; "
+    isValidPayload = traverse parsePact
```

Comment:
confusing name, `isXXX` usually implies a Bool or predicate function. Suggest `validated` or `parsed` or something.

Reply:
I renamed the functions you suggested in recent commit.